lxd/main_forkproxy.go

```diff
@@ -0,0 +1,266 @@
+			fprintf(stderr, "Failed to fork proxy daemon: %s\n", strerror(errno));
+			_exit(1);
+		} else if (childPid != 0) {
+			fprintf(pidFile, "%d", childPid);
+			fclose(pidFile);
+			fclose(stdin);
+			fclose(stdout);
+			fclose(stderr);
+			_exit(0);
+		} else {
+			ret = setsid();
+			if (ret < 0) {
+				fprintf(stderr, "Failed to setsid in proxy daemon: %s\n", strerror(errno));
+				_exit(1);
+			}
+		}
+	}
+
+	// Cannot pass through -1 to runCommand since it is interpreted as a flag
+	fdnum = fdnum == 0 ? -1 : fdnum;
+
+	ret = snprintf(fdpath, sizeof(fdpath), "/proc/self/fd/%d", fdnum);
+	if (ret < 0 || (size_t)ret >= sizeof(fdpath)) {
+		fprintf(stderr, "Failed to format file descriptor path\n");
+		_exit(1);
+	}
+
+	// Join the listener ns if not already setup
+	if (access(fdpath, F_OK) < 0) {
+		// Attach to the network namespace of the listener
+		if (dosetns(listen_pid, "net") < 0) {
+			fprintf(stderr, "Failed setns to listener network namespace: %s\n", strerror(errno));
+			_exit(1);
+		}
+	} else {
+		// Join the connector ns now
+		if (dosetns(connect_pid, "net") < 0) {
+			fprintf(stderr, "Failed setns to connector network namespace: %s\n", strerror(errno));
+			_exit(1);
+		}
+	}
+}
+*/
+import "C"
+
+type cmdForkproxy struct {
+	cmd    *cobra.Command
+	global *cmdGlobal
+}
+
+func (c *cmdForkproxy) Command() *cobra.Command {
+	// Main subcommand
+	cmd := &cobra.Command{}
+	cmd.Use = "forkproxy <listen PID> <listen address> <connect PID> <connect address> <fd> <reexec> <log path> <pid path>"
+	cmd.Short = "Setup network connection proxying"
+	cmd.Long = `Description:
+  Setup network connection proxying
+
+  This internal command will spawn a new proxy process for a particular
+  container, connecting one side to the host and the other to the
+  container.
+`
+	cmd.RunE = c.Run
+	cmd.Hidden = true
+
+	c.cmd = cmd
+	return cmd
+}
+
+func (c *cmdForkproxy) Run(cmd *cobra.Command, args []string) error {
+	if len(args) != 8 {
+		cmd.Help()
+
+		if len(args) == 0 {
+			return nil
+		}
+
+		return fmt.Errorf("Missing required arguments")
+	}
+
+	// Get all our arguments
+	listenPid := args[0]
+	listenAddr := args[1]
+	connectPid := args[2]
+	connectAddr := args[3]
+
+	fd := -1
+	if args[4] != "0" {
+		fd, _ = strconv.Atoi(args[4])
+	}
+
+	// At this point we have already forked and should set this flag to 1
+	args[5] = "1"
+
+	// Check where we are in initialization
+	if !shared.PathExists(fmt.Sprintf("/proc/self/fd/%d", fd)) {
+		fmt.Printf("Listening on %s in %s, forwarding to %s from %s\n", listenAddr, listenPid, connectAddr, connectPid)
+
+		file, err := getListenerFile(listenAddr)
+		if err != nil {
+			return err
+		}
+		defer file.Close()
+
+		listenerFd := file.Fd()
+		if err != nil {
+			return fmt.Errorf("failed to duplicate the listener fd: %v", err)
+		}
+
+		newFd, err := syscall.Dup(int(listenerFd))
+		if err != nil {
+			return fmt.Errorf("failed to dup fd: %v", err)
+		}
+
+		fmt.Printf("Re-executing proxy process\n")
```

Comment:
What is this statement doing exactly?

Reply:
It's effectively just debug logging if you redirect that stuff to a log file I think.